Added explanation for silver milk visits and CPP socdist Solution

TanishT opened the pull request:
I just added my explanation to the current version of the solution.

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
 
+We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 
+
+Graph based off sample test case: <br/>
```
```suggestion
Adjacency list of the sample graph: <br/>
```

frodakcin replied:
Also you should include a photo/diagram of the sample graph.

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+
+Graph based off sample test case: <br/>
+1 (H) --> 2 (H), 5 (G) <br/>
+2 (H) --> 3, 4 (H) <br/>
```
2 also points to 1?

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+Farms 1,2, and 4 are all Holsteins and make up component 1, farm 3 makes up component 2, and farm 5 makes up component 3. 
+
+When it comes to seeing which farmers will be happy, we just have to evalulate 3 conditions: <br/>
+1. Farm $A$ and $B$ are part of the same component. If $C$ is the same breed as $A$ / $B$, then we should output 1.
```
```suggestion
1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output 1.
```

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+
+Connected Components based off of graph: <br/>
+components (first index is 1) = [1, 1, 2, 1, 3] <br/> <br/>
+Farms 1,2, and 4 are all Holsteins and make up component 1, farm 3 makes up component 2, and farm 5 makes up component 3. 
```
Explanation isn't super intuitive. Maybe try something like this?:

```
There are three connected components of the same cow breed in this graph:

- Component 1: Farms 1, 2, and 4
- Component 2: Farm 3
- Component 3: Farm 5

Create an array $\texttt{comp}$ such that $\texttt{comp}[A]$ is the component of farm $A$.

$\texttt{comp} = [1, 1, 2, 1, 3]$ (1-indexed)
```

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+4 (H) --> 2 (H) <br/>
+5 (G) --> 1 (G) <br/>
+
+NOTE: We are not going to store the breed of cow in our adjacency list. The breed is there for convenience.
```
> The breed is there for convenience.

This may confuse people, so just delete it?

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 
+
+Graph based off sample test case: <br/>
+1 (H) --> 2 (H), 5 (G) <br/>
```
I'd just remove the breed completly, it looks sketchy rendered

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+NOTE: We are not going to store the breed of cow in our adjacency list. The breed is there for convenience.
+
+Connected Components based off of graph: <br/>
+components (first index is 1) = [1, 1, 2, 1, 3] <br/> <br/>
```
One indexed instead of first index is 1

HarryWangATX replied:
try latex with variable names

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,40 @@
+components (first index is 1) = [1, 1, 2, 1, 3] <br/> <br/>
+Farms 1,2, and 4 are all Holsteins and make up component 1, farm 3 makes up component 2, and farm 5 makes up component 3. 
+
+When it comes to seeing which farmers will be happy, we just have to evalulate 3 conditions: <br/>
```
We evaluate 3 conditions to check whether the farmer will be happy:

TanishT commented:
I just updated the mdx file according to the changes suggested.

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,46 @@
+4 --> 2 <br/>
+5 --> 1 <br/>
+
+![](<USACO-dec19-milk-visits-silver/graph.png>)
```
can u make this picture more formal lol?

HarryWangATX commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,46 @@
+We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 
+
+Adjacency List based off sample test case: <br/>
+1 --> 2, 5  <br/>
```
can u add latex to this section? personally i think it looks better with latex.

\rightarrow for the arrow btw.

TanishT commented:
I just made the variable names and numbers in my explanation latex. I was kind of confused what you meant by making the graph formal, so I tried to make it as formal as possible (at least in my eyes).

HarryWangATX commented:
is there any way you can make the graph like more straight and text like or maybe even exclude it because idt it helps that much???

TanishT commented:
@HarryWangATX I deleted the graph b/c I thought it was pretty self-explanatory to envision it using the adjacency list explanation. Let me know if there are any other fixes I should make.

bqi343 commented:
Please re-request a review.

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
+Queries for Sample Test Case: 
+1. Farms $1$ and $4$ are in the same component, and since Farmer $1$'s preferred cow is Holstein, they will be satisfied. ($1$)
+2. Farms $1$ and $4$ are in the same component, but since Farmer $2$'s preferred cow is Guernsey, they will be unsatisfied. ($0$)
+3. Farms $1$ and $3$ are in different components, so Farmer $3$ will be satisfied. (1)
```
```suggestion
3. Farms $1$ and $3$ are in different components, so Farmer $3$ will be satisfied. ($1$)
```

to stay consistent with the other queries

frodakcin reviewed (changes requested):
What exactly changed in open20-social-distancing? It was fine as is, no?

TanishT commented:
I just added a cpp solution since there was only a java solution. My cpp solution was also different but used the same binary search for an answer algorithm.

HarryWangATX commented on solutions/USACO open20-social-distancing.mdx:
````diff
@@ -2,15 +2,81 @@
 
 **Time Complexity:** $\mathcal{O}((N+M)\log (maxDist))$
 
+```cpp 
````
why is this outside?

HarryWangATX commented:
> I just added a cpp solution since there was only a java solution. My cpp solution was also different but used the same binary search for an answer algorithm. @frodakcin

wait how is it different? I'm not seeing any difference?? or is it just me??

frodakcin commented:
@TanishT Maybe you forgot to commit and push it? There's no cpp implementation there

TanishT commented:
I deleted it from the file because I didn't want to have commits for 2 different problems in the same or. I can add it if you would like.

frodakcin commented:
Oh, I see. You can make multiple changes in the PR.

frodakcin commented:
If you want you can also rename it

TanishT commented:
Sounds good, I will put my socdist solution into the file. Is there anything else I need to fix for the milkvisits explanation?

frodakcin commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,6 +11,73 @@ author: Qi Wang
+    ll currInterval = 0; //what patch we are currently on
+    ll placed = 0; //how many cows have we placed in total
+    while (placed < n) {
+        if (intervals[currInterval].first <= currPoint && intervals[currInterval].second >= currPoint) {
```
This line has too many characters. (Easiest fix is probably top put a newline after &&)

frodakcin commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,6 +11,73 @@ author: Qi Wang
+        cin >> intervals[i].first >> intervals[i].second;
+    }
+    sort(intervals, intervals + m);
+    ll lo = 1; ll hi = pow(10, 18) + 5;
```
```suggestion
    ll lo = 1; ll hi = (ll)1e18 + 5;
```

Avoid float point

frodakcin replied:
When I write 
```cpp
	ll x = pow(10, 18)+5;
	printf("%lld\n", x);
```
locally it prints out
`1000000000000000000`

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
+
+1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.
+2. Farm $A$ and $B$ are part of the different components. This means that the farmer will always be satisfied because the path between $A$ and $B$ contains both breeds of cows. We should output $1$.
+3. If case $1$ and $2$ are not met, then we should output 0.
```
```suggestion
3. If cases $1$ and $2$ are not met, then we should output 0.
```

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
+
+We evaluate 3 conditions to check whether the farmer will be happy:
+
+1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.
```
```suggestion
1. Farms $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.
```

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
+We evaluate 3 conditions to check whether the farmer will be happy:
+
+1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.
+2. Farm $A$ and $B$ are part of the different components. This means that the farmer will always be satisfied because the path between $A$ and $B$ contains both breeds of cows. We should output $1$.
```
```suggestion
2. Farms $A$ and $B$ are part of different components. This means that the farmer will always be satisfied because the path between $A$ and $B$ contains both breeds of cows. We should output $1$.
```

HarryWangATX commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,6 +11,73 @@ author: Qi Wang
 <LanguageSection>
+    
+<CPPSection>
+
```
```suggestion
## C++ Implementation
```

TanishT commented:
Just updated milkvists and socdist. When I added the ##C++ Implementation, the mdx renderer said that there was an unexpected jsx tag end, and I'm not sure what that means. What should I do to fix the error or should I just delete it?

maggie-j-liu commented:
Try putting a blank line between the `<CPPSection>` tag and the `## C++ Implementation` header like it says [here](https://github.com/cpinitiative/usaco-guide/blob/master/docs/Content%20Documentation.md#using-custom-components).

frodakcin commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,8 +11,80 @@ author: Qi Wang
+    ll placed = 0; //how many cows have we placed in total
+    while (placed < n) {
+        if (intervals[currInterval].first <= currPoint && 
+        intervals[currInterval].second >= currPoint) {
```
```suggestion
            intervals[currInterval].second >= currPoint) {
```

Better formatting

TanishT commented:
> Try putting a blank line between the `<CPPSection>` tag and the `## C++ Implementation` header like it says [here](https://github.com/cpinitiative/usaco-guide/blob/master/docs/Content%20Documentation.md#using-custom-components).

Thanks! It works now.

TanishT commented:
Just changed `ll x = pow(10, 18)+5;` to `ll x = (ll) 1e18 + 5;` and double checked to make sure it prints 1000000000000000005. Anything else I need to do for the milkvists explanation and the socdist c++ solution?

dongliu0426 commented:
change spaces to tabs (https://usaco.guide/general/contributing?lang=cpp#code-conventions)

frodakcin commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,6 +11,76 @@ author: Qi Wang
+    ll placed = 0; //how many cows have we placed in total
+    while (placed < n) {
+        if (intervals[currInterval].first <= currPoint && 
+        intervals[currInterval].second >= currPoint) {
```
```suggestion
            intervals[currInterval].second >= currPoint) {
```
oops

frodakcin replied:
edit: 8 spaces instead of 4... also why does the `if` statement and the second condition line up?

frodakcin replied:
im so confused

frodakcin commented on solutions/USACO open20-social-distancing.mdx:
```diff
@@ -11,139 +11,209 @@ author: Qi Wang
+	ll placed = 0; //how many cows have we placed in total
+	while (placed < n) {
+		if (intervals[currInterval].first <= currPoint && 
+			intervals[currInterval].second >= currPoint) {
```
```suggestion
				intervals[currInterval].second >= currPoint) {
```

frodakcin reviewed (changes requested):
that's the way vim cindent indents

frodakcin reviewed (approved):
If you go [here](https://usaco-guide-aq9df1bqw.vercel.app/problems/) and search for social distancing the link to the binary search module is broken.

frodakcin commented:
Although it does work properly on usaco.guide -- is it just a pr thing?

nwatx commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
 
+We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 
+
+Adjacency List based off sample test case: <br/>
+$1$ $\rightarrow$ $2$, $5$  <br/>
+$2$ $\rightarrow$ $1$, $3$, $4$  <br/>
+$3$ $\rightarrow$ $2$ <br/>
+$4$ $\rightarrow$ $2$ <br/>
+$5$ $\rightarrow$ $1$ <br/>
```
```
Adjacency List based off sample test case:

$1\rightarrow 2,5$

$2\rightarrow 1,3,4$

$3\rightarrow 2$

$4\rightarrow 2$

$5\rightarrow 1$
```
is possibly neater

frodakcin replied:
^ yes do this please

HarryWangATX commented:
> If you go [here](https://usaco-guide-aq9df1bqw.vercel.app/problems/) and search for social distancing the link to the binary search module is broken.

link works for me

HarryWangATX commented:
wait nvm it doesn't oops.

frodakcin commented on solutions/USACO dec19-milkvisits-silver.mdx:
```diff
@@ -2,13 +2,44 @@
 
+We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 
+
+Adjacency List based off sample test case: <br/>
+$1$ $\rightarrow$ $2$, $5$  <br/>
+$2$ $\rightarrow$ $1$, $3$, $4$  <br/>
+$3$ $\rightarrow$ $2$ <br/>
+$4$ $\rightarrow$ $2$ <br/>
+$5$ $\rightarrow$ $1$ <br/>
```
```suggestion
Adjacency List based off sample test case:

$1\rightarrow 2,5$

$2\rightarrow 1,3,4$

$3\rightarrow 2$

$4\rightarrow 2$

$5\rightarrow 1$
```

frodakcin reviewed (approved):
will merge with more approvals
